### PR TITLE
 feat: addListener case

### PR DESCRIPTION
**Pull request by luz33c**
feat：
addListener case
hook components example

### CODE REVIEW

**Comment by EkaanshArora** [comment]
Why is the `addListener` method defined as optional on the `engine`?
Example: `engine.addListener?.('onLocalVideoStateChanged',...)`

This should be a static method and **not** optional.

**Comment by EkaanshArora** [comment]
@LichKing-2234 can you take a look

**Comment by LichKing-2234** [comment]
@EkaanshArora 
If not defined as optional, the method must be implemented in `IAgoraRtcEngineImpl.ts`, but it is generated file, we just want to implement it in `RtcEngineExInternal.ts`. Any suggestions?

**Review by LichKing-2234 — APPROVED** [review]

**Comment by EkaanshArora** [comment]
Can we modify the way the file is generated? This can be a blocker for adoption.

**Comment by LichKing-2234** [comment]
> Can we modify the way the file is generated? This can be a blocker for adoption.

@EkaanshArora if disable strict mode, can keep using `engine.addListener('onLocalVideoStateChanged',...)`, otherwise can use `engine.addListener?('onLocalVideoStateChanged',...)` or `engine.addListener!('onLocalVideoStateChanged',...)`

**Comment by EkaanshArora** [comment]
@LichKing-2234 This isn't acceptable. We cannot ask developers to disable strict mode, this takes away all benefits of using Typescript.

**Comment by luz33c** [comment]
> @LichKing-2234 This isn't acceptable. We cannot ask developers to disable strict mode, this takes away all benefits of using Typescript.

I will fix your problem, just define an interface for the engine

File: example/src/examples/hooks/JoinChannelVideo/JoinChannelVideo.tsx [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+          'onUserOffline',
+          'connection',
+          connection,
+          'remoteUid',
+          remoteUid,
+          'reason',
+          reason
+        );
+        if (remoteUsers === undefined) return;
+        setRemoteUsers([...remoteUsers!, remoteUid]);
```
Comment (EkaanshArora) [inline example/src/examples/hooks/JoinChannelVideo/JoinChannelVideo.tsx:206]: The state should remove the user that has left: 
```tsx
setRemoteUsers((prev) => prev.filter((uid) => uid !== remoteUid));
````

**Review by EkaanshArora — COMMENTED** [review]
The example has some issues.

File: example/src/examples/hooks/JoinChannelVideo/JoinChannelVideo.tsx (hunk 1 above)
Comment (EkaanshArora) [reply example/src/examples/hooks/JoinChannelVideo/JoinChannelVideo.tsx:206]: I can submit a PR to fix the example if needed.

**Pull request merged**